pkg/handler/common/provider/kubevirt.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+
+	apiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	apiv2 "k8c.io/kubermatic/v2/pkg/api/v2"
+	handlercommon "k8c.io/kubermatic/v2/pkg/handler/common"
+	"k8c.io/kubermatic/v2/pkg/handler/middleware"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/provider/cloud/kubevirt"
+	kubevirtcli "k8c.io/kubermatic/v2/pkg/provider/cloud/kubevirt/kubevirtcli/client/versioned"
+	kubernetesprovider "k8c.io/kubermatic/v2/pkg/provider/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/util/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+)
+
+const (
+	// soon the namespace will be changed to the cluster namespace
+	kubevirtns = "default"
+)
+
+var NewKubevirtClientSet = func(kubeconfig string) (kubevirtcli.Interface, kubernetes.Interface, error) {
+	config, err := base64.StdEncoding.DecodeString(kubeconfig)
+	if err != nil {
+		// if the decoding failed, the kubeconfig is sent already decoded without the need of decoding it,
+		// for example the value has been read from Vault during the ci tests, which is saved as json format.
+		config = []byte(kubeconfig)
```

Comment:
what if you get a malformed kubeconfig? In your case, you can't recognize it. You can check if the kubeconfig is base64 and then decode and throw the error if happened
```
func isBase64(input string) bool {
	r, _ := regexp.Compile(`^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)?$`)
	return r.Match([]byte(line))
}
```

Reply:
Second thought. You send kubeconfig in the header in the endpoint. I think it always must be base64. Please verify this. We also use base64 for sending kubeconfig in other endpoints, if true then `config = []byte(kubeconfig)` it's not necessary

Reply:
It's true that it will always be base64 encoded. I have updated accordingly.